### PR TITLE
Added address validation wizard to Address model. #6733

### PR DESCRIPTION
**Pull request by PritishC**

### CODE REVIEW

**Comment by PritishC** [comment]
I tested out the validation wizard by installing my custom shipping module in a UPS database, so as to have a default configuration provider.

addrvalupsstart

addrvalupsstart_2

File: party.py [hunk 1]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
-        return getattr(self, '_{0}_address_validate'.format(provider))()
+        return getattr(
+            self, '_{0}_address_validate'.format(provider)
+        )()  # pragma: no cover
+
+    def serialize(self):
+        """
+        Serialize address record.
+        """
+        return {
```
Comment (tarunbhardwaj) [inline party.py]: use purpose

File: party.py [hunk 2]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
+
+
+class AddressValidationDiffView(ModelView):
+    """
+    Address Validation Wizard Diff View
+    """
+    __name__ = 'party.address.validation.start'
+
+    # Older Values
+    old_name = fields.Char('Current Name', readonly=True)
```
Comment (tarunbhardwaj) [inline party.py]: current_name

File: party.py [hunk 3]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
+    old_city = fields.Char('Current City', readonly=True)
+    old_country = fields.Many2One(
+        'country.country', 'Current Country', readonly=True
+    )
+    old_subdivision = fields.Many2One(
+        "country.subdivision", 'Current Subdivision', readonly=True
+    )
+
+    # Editable suggestions
+    name = fields.Char('Suggested Name')
```
Comment (tarunbhardwaj) [inline party.py]: no need for editable name

File: party.py [hunk 4]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
+    street = fields.Char('Suggested Street')
+    zip = fields.Char('Suggested Zip')
+    city = fields.Char('Suggested City')
+    country = fields.Many2One('country.country', 'Suggested Country')
+    subdivision = fields.Many2One(
```
Comment (tarunbhardwaj) [inline party.py]: domain=[('country', '=', Eval('country'))]

File: party.py [hunk 5]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
+        "country.subdivision", 'Suggested Subdivision',
+        domain=[('country', '=', Eval('country'))],
+        depends=['country']
+    )
+
+
+class AddressValidationDone(ModelView):
```
Comment (tarunbhardwaj) [inline party.py]: AddressValidationMsg

File: party.py [hunk 6]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
+
+    done_msg = fields.Text("Message", readonly=True)
+
+    @staticmethod
+    def default_done_msg():  # pragma: no cover
+        """
+        Returns default message.
+        """
+        return (
+            'Address validation is complete.'
```
Comment (tarunbhardwaj) [inline party.py]: can we add full_address string to this message

File: party.py [hunk 7]
```diff
@@ -93,4 +114,168 @@ def _providername_address_validate(self):
...
+            'name': self.name or None,
+            'street': self.street or None,
+            'zip': self.zip or None,
+            'city': self.city or None,
+            'country': self.country and self.country.id or None,
+            'subdivision': self.subdivision and self.subdivision.id or None,
+        }
+
+
+class AddressValidationDiffView(ModelView):
```
Comment (tarunbhardwaj) [inline party.py]: AddressValidationSuggestionView

File: party.py (hunk 6 above)
Comment (PritishC) [inline party.py]: Done in `default_done()` method in the wizard.

File: party.py [hunk 8]
```diff
@@ -93,4 +114,177 @@ def _providername_address_validate(self):
...
+        """
+        if purpose == 'validation':
+            return {
+                'name': self.name or None,
+                'street': self.street or None,
+                'zip': self.zip or None,
+                'city': self.city or None,
+                'country': self.country and self.country.id or None,
+                'subdivision': self.subdivision and self.subdivision.id or None,
+            }
```
Comment (tarunbhardwaj) [inline party.py]: else call super

File: party.py [hunk 9]
```diff
@@ -93,4 +114,177 @@ def _providername_address_validate(self):
...
+
+
+class AddressValidationSuggestionView(ModelView):
+    """
+    Address Validation Wizard Diff View
```
Comment (tarunbhardwaj) [inline party.py]: Suggestion

File: party.xml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        <record model="ir.action.wizard" id="wizard_address_validation">
+            <field name="name">Address Validation</field>
+            <field name="wiz_name">party.address.validation</field>
+            <field name="model">party.address</field>
+        </record>
+
+        <record model="ir.ui.view" id="address_diff_view_form">
+            <field name="model">party.address.validation.start</field>
+            <field name="type">form</field>
+            <field name="name">address_diff_form</field>
```
Comment (tarunbhardwaj) [inline party.xml]: address_suggestion_form

File: party.xml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+            <field name="name">address_form</field>
+        </record>
+
+        <record model="ir.action.wizard" id="wizard_address_validation">
+            <field name="name">Address Validation</field>
+            <field name="wiz_name">party.address.validation</field>
+            <field name="model">party.address</field>
+        </record>
+
+        <record model="ir.ui.view" id="address_diff_view_form">
```
Comment (tarunbhardwaj) [inline party.xml]: address_suggestion_view_form

File: party.xml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+            <field name="model">party.address.validation.start</field>
+            <field name="type">form</field>
+            <field name="name">address_diff_form</field>
+        </record>
+
+        <record model="ir.ui.view" id="address_end_view_form">
+            <field name="model">party.address.validation.end</field>
+            <field name="type">form</field>
+            <field name="name">address_end_form</field>
```
Comment (tarunbhardwaj) [inline party.xml]: Address_validation_end_form

File: party.xml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+            <field name="model">party.address</field>
+        </record>
+
+        <record model="ir.ui.view" id="address_diff_view_form">
+            <field name="model">party.address.validation.start</field>
+            <field name="type">form</field>
+            <field name="name">address_diff_form</field>
+        </record>
+
+        <record model="ir.ui.view" id="address_end_view_form">
```
Comment (tarunbhardwaj) [inline party.xml]: Address_validation_end_view_form

File: view/address_diff_form.xml [hunk 14]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0"?>
+<form string="Address Validation Form" col="6">
```
Comment (tarunbhardwaj) [inline view/address_diff_form.xml]: Address validation suggestion form

File: view/address_end_form.xml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0"?>
+<form string="Address Validation Complete">
+	<field name="done_msg"/>
```
Comment (tarunbhardwaj) [inline view/address_end_form.xml]: fix indentation

File: view/address_form.xml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<data>
+	<xpath expr="/form/field[@name='subdivision']" position="after">
```
Comment (tarunbhardwaj) [inline view/address_form.xml]: fix indentation

File: view/address_diff_form.xml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+	<label name="current_name"/>
```
Comment (tarunbhardwaj) [inline view/address_diff_form.xml]: indentation

File: party.py (hunk 8 above)
Comment (PritishC) [inline party.py]: Done.

**Pull request merged**